### PR TITLE
Enrollment Client Library

### PR DESCRIPTION
**Pull request by nicolaspapp**
|Related issue|
|---|
| #4900 |

## Description

Define a library that handles the enrollment process from the agent side. Should replace the main code in `agent-auth` and will later be used for an auto-registration process
Current defined interfaces:

```
/**
 * @brief Struct that defines the connection target
 * */
typedef struct _enrollment_target_cfg {
    char *manager_name;        /**> Manager's direction or ip address */
    int port;                 /**> Manager's port                     */
    char *agent_name;         /**> (optional) Name of the agent. In case of NULL enrollment message will send local hostname */
    char *centralized_group;  /**> (optional) In case the agent belong to a group */
    char *sender_ip;          /**> (optional) IP adress or CIDR of the agent. In case of null the manager will use the source ip */
} w_enrollment_target;

/**
 * @brief Certificate configurations 
 * 
 * Struct that defines the enrollment certificate configuration
 * Client Enrollment methods:
 * 1. Simple verification (only chipers needed)
 * 2. Password (uses authpass param)
 * 3. Manager Verificatiion (uses ca_cert param) 
 * 4. Manager and Agent Verification (uses agent_cert and agent_key params)
 */
typedef struct _enrollment_cert_cfg {
    char *ciphers;     /**> chipers string (default DEFAULT_CIPHERS) */
    char *authpass;    /**> for password verification */
    char *agent_cert;  /**> Agent Certificate (null if not used) */
    char *agent_key;   /**> Agent Key (null if not used) */
    char *ca_cert;     /**> CA Certificate to verificate server (null if not used) */
    int auto_method:1; /**> 0 for TLS v1.2 only (Default), 1 for Auto negotiate the most secure common SSL/TLS method with the client. */
} w_enrollment_cert; 

/**
 * @brief Strcture that handles all the enrollment configuration
 * */
typedef struct _enrollment_ctx {
    const w_enrollment_target *target_cfg;  /**> for details @see _enrollment_target_cfg */
    const w_enrollment_cert *cert_cfg;      /**> for details @see _enrollment_cert_cfg */
    SSL *ssl;                                   /**> will hold the connection instance with the manager */
    unsigned int enabled:1;
} w_enrollment_ctx;

/**
 * Initializes parameters of an w_enrollment_ctx structure based
 * on a target and certificate configurations
 * */
w_enrollment_ctx * w_enrollment_init(const w_enrollment_target *target, const w_enrollment_cert *cert);

/**
 * Frees parameers of an w_enrollment_ctx structure
 * target_cfg and cert_cfg should be freed on their own since there are constant pointers
 * */
void w_enrollment_destroy(w_enrollment_ctx *cfg);

/**
 * @brief Generates an enrollment process
 * 
 * @param cfg configuration @see w_enrollment_ctx
 * @param server_adress (optional) If null server_adress will be obtained from cfg
 * */
int w_enrollment_request_key(w_enrollment_ctx *cfg, const char * server_address);
```

## Tests

- Manual test for available enrollment methods:
  1. Simple verification
  2. Simple verification with Password
  3. Manger Verificatiion (CA certificates)  
  4. Manger and Agent Verification (SSL Certificate signed by manager)
- Unit Test for all library methods

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
  - [x] MAC OS X
- [x] Review logs syntax and correct language

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)
- Memory tests for Windows
  - [x] Scan-build report
  - [x] Coverity
- Memory tests for macOS
  - [ ] Scan-build report
  - [ ] Leaks
  - [ ] AddressSanitizer

- [x] Retrocompatibility with older Wazuh versions
- [x] The data flow works as expected (agent-manager-api-app)
- [x] Added unit tests (for new features)

### CODE REVIEW

**Comment by nicolaspapp** [comment]
## Valgrind run
36 Unit Tests passing without leaks

```
==18569== Memcheck, a memory error detector
==18569== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==18569== Using Valgrind-3.12.0.SVN and LibVEX; rerun with -h for copyright info
==18569== Command: shared/test_enrollment_op
==18569== 
[==========] Running 36 test(s).
[ RUN      ] test_w_enrollment_concat_src_ip_invalid_ip
[       OK ] test_w_enrollment_concat_src_ip_invalid_ip
[ RUN      ] test_w_enrollment_concat_src_ip_valid_ip
[       OK ] test_w_enrollment_concat_src_ip_valid_ip
[ RUN      ] test_w_enrollment_concat_src_ip_empty_ip
[       OK ] test_w_enrollment_concat_src_ip_empty_ip
[ RUN      ] test_w_enrollment_concat_src_ip_empty_buff
Expected assertion buff != NULL occurred
[       OK ] test_w_enrollment_concat_src_ip_empty_buff
[ RUN      ] test_w_enrollment_concat_group_empty_buff
Expected assertion buff != NULL occurred
[       OK ] test_w_enrollment_concat_group_empty_buff
[ RUN      ] test_w_enrollment_concat_group_empty_group
Expected assertion centralized_group != NULL occurred
[       OK ] test_w_enrollment_concat_group_empty_group
[ RUN      ] test_w_enrollment_concat_group
[       OK ] test_w_enrollment_concat_group
[ RUN      ] test_w_enrollment_verify_ca_certificate_null_connection
Expected assertion ssl != NULL occurred
[       OK ] test_w_enrollment_verify_ca_certificate_null_connection
[ RUN      ] test_w_enrollment_verify_ca_certificate_no_certificate
[       OK ] test_w_enrollment_verify_ca_certificate_no_certificate
[ RUN      ] test_verificy_ca_certificate_invalid_certificate
[       OK ] test_verificy_ca_certificate_invalid_certificate
[ RUN      ] test_verificy_ca_certificate_valid_certificate
[       OK ] test_verificy_ca_certificate_valid_certificate
[ RUN      ] test_w_enrollment_connect_empty_address
Expected assertion server_address != NULL occurred
[       OK ] test_w_enrollment_connect_empty_address
[ RUN      ] test_w_enrollment_connect_empty_config
Expected assertion cfg != NULL occurred
[       OK ] test_w_enrollment_connect_empty_config
[ RUN      ] test_w_enrollment_connect_invalid_hostname
[       OK ] test_w_enrollment_connect_invalid_hostname
[ RUN      ] test_w_enrollment_connect_could_not_setup
[       OK ] test_w_enrollment_connect_could_not_setup
[ RUN      ] test_w_enrollment_connect_socket_error
[       OK ] test_w_enrollment_connect_socket_error
[ RUN      ] test_w_enrollment_connect_SSL_connect_error
[       OK ] test_w_enrollment_connect_SSL_connect_error
[ RUN      ] test_w_enrollment_connect_success
[       OK ] test_w_enrollment_connect_success
[ RUN      ] test_w_enrollment_send_message_empty_config
Expected assertion cfg != NULL occurred
[       OK ] test_w_enrollment_send_message_empty_config
[ RUN      ] test_w_enrollment_send_message_wrong_hostname
[       OK ] test_w_enrollment_send_message_wrong_hostname
[ RUN      ] test_w_enrollment_send_message_ssl_error
[       OK ] test_w_enrollment_send_message_ssl_error
[ RUN      ] test_w_enrollment_send_message_success
[       OK ] test_w_enrollment_send_message_success
[ RUN      ] test_w_enrollment_store_key_entry_null_key
Expected assertion keys != NULL occurred
[       OK ] test_w_enrollment_store_key_entry_null_key
[ RUN      ] test_w_enrollment_store_key_entry_cannot_open
[       OK ] test_w_enrollment_store_key_entry_cannot_open
[ RUN      ] test_w_enrollment_store_key_entry_success
[       OK ] test_w_enrollment_store_key_entry_success
[ RUN      ] test_w_enrollment_process_agent_key_empty_buff
Expected assertion buffer != NULL occurred
[       OK ] test_w_enrollment_process_agent_key_empty_buff
[ RUN      ] test_w_enrollment_process_agent_key_short_buff
Expected assertion strlen(buffer) > 9 occurred
[       OK ] test_w_enrollment_process_agent_key_short_buff
[ RUN      ] test_w_enrollment_process_agent_key_invalid_format
[       OK ] test_w_enrollment_process_agent_key_invalid_format
[ RUN      ] test_w_enrollment_process_agent_key_invalid_key
[       OK ] test_w_enrollment_process_agent_key_invalid_key
[ RUN      ] test_w_enrollment_process_agent_key_valid_key
[       OK ] test_w_enrollment_process_agent_key_valid_key
[ RUN      ] test_w_enrollment_process_response_ssl_null
Expected assertion ssl != NULL occurred
[       OK ] test_w_enrollment_process_response_ssl_null
[ RUN      ] test_w_enrollment_process_response_ssl_error
[       OK ] test_w_enrollment_process_response_ssl_error
[ RUN      ] test_w_enrollment_process_response_message_error
[       OK ] test_w_enrollment_process_response_message_error
[ RUN      ] test_w_enrollment_process_response_success
[       OK ] test_w_enrollment_process_response_success
[ RUN      ] test_w_enrollment_request_key_null_cfg
Expected assertion cfg != NULL occurred
[       OK ] test_w_enrollment_request_key_null_cfg
[ RUN      ] test_w_enrollment_request_key
[       OK ] test_w_enrollment_request_key
[==========] 36 test(s) run.
[  PASSED  ] 36 test(s).
==18569== 
==18569== HEAP SUMMARY:
==18569==     in use at exit: 0 bytes in 0 blocks
==18569==   total heap usage: 5,836 allocs, 5,836 frees, 1,653,752 bytes allocated
==18569== 
==18569== All heap blocks were freed -- no leaks are possible
==18569== 
==18569== For counts of detected and suppressed errors, rerun with: -v
==18569== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by nicolaspapp** [comment]
## Coverage report 

96.7% of coverage

image

File: src/shared/enrollment_op.c [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+static int w_enrollment_store_key_entry(const char* keys);
+
+/* Constants */
+static const int ENTRY_ID = 0;
+static const int ENTRY_NAME = 1;
+static const int ENTRY_IP = 2;
+static const int ENTRY_KEY = 3; 
+
+w_enrollment_ctx * w_enrollment_init(const w_enrollment_target *target, const w_enrollment_cert *cert) {
+    w_enrollment_ctx *cfg;
```
[the remark this answers is not in the document]
Comment (nicolaspapp) [reply src/shared/enrollment_op.c:52]: Agreed, I seem to have forgotten the assertions in both init an destroy methods

File: src/os_auth/main-client.c [hunk 2]
```diff
@@ -133,7 +128,7 @@ int main(int argc, char **argv)
                 if (!optarg) {
                     merror_exit("-g needs an argument");
                 }
-                dir = optarg;
+                // dir = optarg; (NEVER USED)
```
Comment (palaciosjeremias) [inline src/os_auth/main-client.c:131]: If this option is no longer used, we can add the following message: mwarn(DEPRECATED_OPTION_WARN,"-D");

**Pull request merged**